File: src/main/java/alg01unionfind/UF1QuickFind.java
package alg01unionfind;


/**
 * <p>
 * Defects:
 * <ul>
 * <li>Union is expensive</li>
 * </ul>
 * </p>
 */
public class UF1QuickFind implements UnionFindAlgorithm {

    private int[] id;

    public UF1QuickFind(int size) {
        id = new int[size];
        for (int i = 0; i < size; i++) {
            id[i] = i;
        }
    }

    /**
     * This is fast as O(1), hence "QuickFind" name
     */
    public boolean connected(int n1, int n2) {
        return id[n1] == id[n2];
    }

    /**
     * But this is slow as O(N)
     */
    public void union(int n1, int n2) {
        int n1id = id[n1];
        int n2id = id[n2];
        // We change every occurrence of n1id to n2id
        // And this is bad, because it grows as O(N)
        // This will be ultra slow as O(N^2) if we will have
        // N UnionFinds
        for (int i = 0; i < id.length; i++) {
            if (id[i] == n1id) {
                id[i] = n2id;
            }
        }
    }

}

File: src/main/kotlin/alg01unionfind/UF1QuickFind.kt
package alg01unionfind

/**
 * Find is quick, but union is slow: hence the name.
 */
class UF1QuickFind(size: Int) : UnionFind {
    private val ids: Array<Int> = Array(size) { index: Int -> index }

    /**
     * This is slow as O(N).
     */
    override fun union(n1: Int, n2: Int) {
        val n1id = this.ids[n1]
        val n2id = this.ids[n2]
        /*
        Every occurrence of n1id is changed to n2id.
        This is as slow as O(N) and is going to be O(N^2) if N union find is required.
         */
        for (index in this.ids.indices) {
            if (this.ids[index] == n1id) {
                this.ids[index] = n2id
            }
        }
    }

    /**
     * This is fast as O(1).
     */
    override fun connected(n1: Int, n2: Int): Boolean {
        return ids[n1] == ids[n2]
    }
}